File: ds/union-find.py
# https://atcoder.jp/contests/abc075/tasks/abc075_c
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_1.cpp
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_4.cpp

class UnionFind:
    def __init__(self, N):
        self.parent = [-1] * N
        self._size = [1] * N

    def root(self, x):
        if self.parent[x] == -1:
            return x;
        else:
            ans = self.root(self.parent[x])
            self.parent[x] = self.root(self.parent[x])
            return ans

    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return False
        else:
            if self._size[x] < self._size[y]:
                x, y = y, x  # Union by size
            self.parent[y] = x
            self._size[x] += self._size[y]
            return True

    def size(self, x):
        return self._size[self.root(x)]


N, M = map(int, input().split())
graph = []
for m in range(M):
    a, b = map(int, input().split())
    graph.append((a - 1, b - 1))

ans = 0
for m in range(M):
    uf = UnionFind(N)
    for e in range(M):
        if m == e:
            continue
        uf.unite(graph[e][0], graph[e][1])
    count = 0
    for n in range(N):
        if uf.root(n) == n:
            count += 1
    if count == 2:
        ans += 1

print(ans)

File: ds/union-find.cc
// https://atcoder.jp/contests/abc229/tasks/abc229_e
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_1.cpp
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_4.cpp

#include <iostream>
#include <vector>
using namespace std;

struct UnionFind {
  vector<int> parent, _size;

  UnionFind(int n) : parent(n, -1), _size(n, 1) {}

  int root(int x) {
    if (parent[x] == -1) {
      return x;
    } else {
      //return root(parent[x]);  // O(logN)
      return parent[x] = root(parent[x]);  // O(alpha(N))
    }
  }

  bool issame(int x, int y) {
    return root(x) == root(y);
  }

  bool unite(int x, int y) {
    x = root(x);
    y = root(y);
    if (x == y) {
      return false;
    }
    if (_size[x] < _size[y]) {
      swap(x, y);  // Union by size
    }
    parent[y] = x;
    _size[x] += _size[y];
    return true;
  }

  int size(int x) {
    return _size[root(x)];
  }
};

int main() {
  int N, M;
  cin >> N >> M;
  vector<vector<int> > graph(N);
  for (int m = 0; m < M; m++) {
    int a, b;
    cin >> a >> b;
    a--;
    b--;
    graph[a].push_back(b);
  }
  vector<int> ans(N, 0);
  UnionFind uf(N);
  int num = 0;  // Union-Findの連結成分数
  for (int n = N - 1; n > 0; n--) {
    num++;
    for (size_t i = 0; i < graph[n].size(); i++) {
      int u = graph[n][i];
      if (uf.unite(n, u)) {
        num--;
      }
    }
    ans[n - 1] = num;
  }
  for (int n = 0; n < N; n++) {
    cout << ans[n] << endl;
  }
  return 0;
}
